gym_xiangqi/envs/xiangqi_env.py

```diff
@@ -21,5 +178,278 @@ def render(self, mode='human'):
+        """
+        Find legal moves for the Chariots
+        """
+        for pid in range(CHARIOT_1, CHARIOT_2+1):
+            orig_pos = (self.agent_piece[pid].row,
+                        self.agent_piece[pid].col)
+
+            for offset in ORTHOGONAL:
+                next_pos = (orig_pos[0] + offset[0], orig_pos[1] + offset[1])
+                # No need to check for repetition; check as far as possible
+                self.check_action(pid, orig_pos, next_pos, MAX_REP, offset, 0)
+
+    def get_cannon_actions(self):
+        """
+        Find legal moves for the Cannons
+        """
+        for pid in range(CANNON_1, CANNON_2+1):
+            orig_pos = (self.agent_piece[pid].row,
+                        self.agent_piece[pid].col)
+
+            for offset in ORTHOGONAL:
+                # moving positions
+                next_pos = (orig_pos[0] + offset[0], orig_pos[1] + offset[1])
+                reps = self.check_action(pid, orig_pos, next_pos,
+                                         MAX_REP, offset, 0)
+
+                # mark the farthest position invalid if it is an enemy
+                last_r = orig_pos[0] + offset[0] * reps
+                last_c = orig_pos[1] + offset[1] * reps
+
+                if self.state[last_r][last_c] < 0:
+                    action_idx = self.move_to_action_space(pid, orig_pos,
+                                                           (last_r, last_c))
+                    self.possible_actions[action_idx] = 0
+
+                # attacking positions
+                next_r = orig_pos[0] + offset[0] * (reps + 1)
+                next_c = orig_pos[1] + offset[1] * (reps + 1)
+
+                while True:
+                    row_bound = 0 <= next_r < self.shape[0]
+                    col_bound = 0 <= next_c < self.shape[1]
+
+                    if not row_bound or not col_bound:
+                        break
+
+                    if self.state[next_r][next_c] > 0:
+                        break
+                    elif self.state[next_r][next_c] < 0:
+                        action_idx = self.move_to_action_space(
+                            pid, orig_pos, (next_r, next_c)
+                        )
+                        self.possible_actions[action_idx] = 1
+                        break
+
+                    next_r += offset[0]
+                    next_c += offset[1]
+
+    def get_soldier_actions(self):
+        """
+        Find legal moves for the soldiers
+        """
+        for pid in range(SOLDIER_1, SOLDIER_5+1):
+            orig_pos = (self.agent_piece[pid].row,
+                        self.agent_piece[pid].col)
+
+            if orig_pos[0] > 4:
+                moves = [0]
+            else:
+                moves = [0, 1, 3]
+
+            for i in moves:
+                offset = ORTHOGONAL[i]
+                next_pos = (orig_pos[0] + offset[0], orig_pos[1] + offset[1])
+                self.check_action(pid, orig_pos, next_pos, 1, offset, 0)
```

Comment:
Not sure where to put this, but there's an empty file at gym_xiangqi/test/__init__.py. Is that for some specific reason?

Reply:
deleted __init__.py